pkg/ingestor/parser/vuln/vuln.go

```diff
@@ -0,0 +1,167 @@
+// attestation document.  The attestation node is linked to the package node
+// via a "attestation" edge type.
+//
+// Depending on the number of vulnerabilities found, vulnerability nodes are
+// generated that contain just the vulnerability ID that can be used to query
+// for more information as needed.  The vulnerability node is linked to the
+// attestation node via a "vulnerable" edge type.
+package vuln
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/guacsec/guac/pkg/assembler"
+	"github.com/guacsec/guac/pkg/assembler/clients/generated"
+	"github.com/guacsec/guac/pkg/assembler/helpers"
+	attestation_vuln "github.com/guacsec/guac/pkg/certifier/attestation"
+	"github.com/guacsec/guac/pkg/handler/processor"
+	"github.com/guacsec/guac/pkg/ingestor/parser/common"
+)
+
+type parser struct {
+	packages []*generated.PkgInputSpec
+	vulnData *generated.VulnerabilityMetaDataInput
+	vulns    []*generated.OSVInputSpec
+	isVulns  []assembler.IsVulnIngest
+}
+
+// NewVulnCertificationParser initializes the parser
+func NewVulnCertificationParser() common.DocumentParser {
+	return &parser{}
+}
+
+// Parse breaks out the document into the graph components
+func (c *parser) Parse(ctx context.Context, doc *processor.Document) error {
+	statement, err := parseVulnCertifyPredicate(doc.Blob)
+	if err != nil {
+		return fmt.Errorf("failed to parse slsa predicate: %w", err)
+	}
+	ps, err := parseSubject(statement)
+	if err != nil {
+		return fmt.Errorf("unable to parse subject of statement: %w", err)
+	}
+	c.packages = ps
+	c.vulnData = parseMetadata(statement, doc)
+	vs, ivs, err := parseVulns(statement, doc)
+	if err != nil {
+		return fmt.Errorf("unable to parse vulns of statement: %w", err)
+	}
+	c.vulns = vs
+	c.isVulns = ivs
+	return nil
+}
+
+func parseVulnCertifyPredicate(p []byte) (*attestation_vuln.VulnerabilityStatement,
+	error) {
+	predicate := attestation_vuln.VulnerabilityStatement{}
+	if err := json.Unmarshal(p, &predicate); err != nil {
+		return nil, err
+	}
+	return &predicate, nil
+}
+
+func parseSubject(s *attestation_vuln.VulnerabilityStatement) ([]*generated.PkgInputSpec,
+	error) {
+	var ps []*generated.PkgInputSpec
+	for _, sub := range s.StatementHeader.Subject {
+		p, err := helpers.PurlToPkg(sub.Name)
+		if err != nil {
+			return nil, fmt.Errorf("bad purl in statement header: %w", err)
+		}
+		ps = append(ps, p)
+		// FIXME Do we do anything with Digest here?
+	}
+	return ps, nil
+}
+
+func parseMetadata(s *attestation_vuln.VulnerabilityStatement,
+	doc *processor.Document) *generated.VulnerabilityMetaDataInput {
+	return &generated.VulnerabilityMetaDataInput{
+		TimeScanned:    *s.Predicate.Metadata.ScannedOn,
+		DbUri:          s.Predicate.Scanner.Database.Uri,
+		DbVersion:      s.Predicate.Scanner.Database.Version,
+		ScannerUri:     s.Predicate.Scanner.Uri,
+		ScannerVersion: s.Predicate.Scanner.Version,
+		Origin:         doc.SourceInformation.Source,
+		Collector:      doc.SourceInformation.Collector,
```

Comment:
the origin and collector stuff will be handled by the [graph builder](https://github.com/guacsec/guac/blob/main/pkg/ingestor/parser/common/graph_builder.go), not required to set inside the parser, also will need to add something [here](https://github.com/guacsec/guac/blob/main/pkg/ingestor/parser/common/graph_builder.go#L61).

(it looks like i may have forgotten to add plumbing for Slsa predicate there, if you want to help with a drive-by :) )

Reply:
Got it, I'll add this and slsa there.

Reply:
Done.